src/uct/ib/rdmacm/rdmacm_cm.h

```diff
@@ -50,6 +56,36 @@ typedef struct uct_rdmacm_cm_config {
     double          timeout;
 } uct_rdmacm_cm_config_t;
 
+struct uct_rdmacm_cm_reserved_qpn_blk {
+    ucs_bitmap_t(256)       bmp;
+    uint32_t                first_qpn;
+    unsigned int            next_avail_qpn_offset;
+    ucs_list_link_t         entry;
+    struct mlx5dv_devx_obj  *obj;
+};
+
+struct uct_rdmacm_cm_reserved_qpn_context {
+     /* Avoid competed condition on the qpn resource for multi-threads */ 
+     pthread_mutex_t mutex;
+     ucs_list_link_t blk_list;
+     bool            support;
+     uint32_t        log_reserved_qpn_granularity;
+     uint64_t        general_obj_types_caps;
+ };
+
+struct uct_rdmacm_cm_dummy_qp_context {
+    struct ibv_cq *cq;
+};
+
+struct uct_rdmacm_cm_qp_context {
+    bool reserved_qpn_support;
+
+    union {
+        struct uct_rdmacm_cm_reserved_qpn_context reserved_qpn_ctx;
+        struct uct_rdmacm_cm_dummy_qp_context     dummy_qp_ctx;
+    };
```

Comment:
union here makes code more complicated just to save 8 bytes per RDMA device, it's not worth it
pls remove union and unwrap all nested structs inside this structs

Reply:
@yosefe The union was added according to below comment. Since we already changed to use one hash, then maybe better to keep the union, otherwise when the qp is dummy qp then without union it will cost more bytes(the size of qpn context).

https://github.com/openucx/ucx/pull/6604#discussion_r621938890

Reply:
> @yosefe The union was added according to below comment. Since we already changed to use one hash, then maybe better to keep the union, otherwise when the qp is dummy qp then without union it will cost more bytes(the size of qpn context).
> 
> [#6604 (comment)](https://github.com/openucx/ucx/pull/6604#discussion_r621938890)

@yosefe My mistake,  since we are using one hash, so yes, without union it will always cost 8 more bytes as you mentioned.

Reply:
since hash is per device, this extra memory consumption is not an issue

Reply:
@yosefe Fixed.